### DISCUSSION TITLE
ThingSpeak : Different ThingSpeak APIs

### DISCUSSION THREAD

**Discussion by yesoreyeram** [grafana/grafana-infinity-datasource#38] [Specific APIs]
In this example, we are going to visualise data from ThingSpeak API. 

For demo purpose, I am using [traffic monitor public channel](https://thingspeak.com/channels/38629) ( Channel ID : 38629 ). You may know already that, ThingSpeak APIs available in different formats such as JSON, CSV and XML. We are going to see how to use all of those different API formats.

# JSON Example 

As you can see in the below picture, I am using the [JSON feed url](https://thingspeak.com/channels/38629/feed.json) of the channel. If you look at the JSON schema, my data points are under the node called `feeds`. So I specify the same as my `Rows/Root` selector. Then for the columns, we need to first specify our time column which is `created_at` field in our json feed. Then we can specify one or more metrics(number) fields as shown in the below picture.

![image](https://user-images.githubusercontent.com/153843/108479371-9030bb00-728d-11eb-8ae5-f186c78db64e.png)

# CSV Example

If you prefer CSV format over JSON, here is the example settings of the same channel. I am using [CSV feed url](https://thingspeak.com/channels/38629/feed.csv) of the channel. In the query, Choose `CSV` as your type, specify the csv feed URL and then the fields you needed along with time field.

![image](https://user-images.githubusercontent.com/153843/108479976-4b595400-728e-11eb-868e-b2d550f496f3.png)

# XML Example

You are not limited to CSV, JSON. You can also use xml format as well in your query if your prefer that. I am using [XML feed url](https://thingspeak.com/channels/38629/feed.xml) of the channel. Query setup is almost same as csv/json. Notable configurations are your rows/root field and timestamp field.   As you see in the below picture, I am using `channel.feeds[0].feed` as my root/rows selector and then for selecting timestamp, I use `created-at[0]._`

![image](https://user-images.githubusercontent.com/153843/108480329-b99e1680-728e-11eb-91f3-38c5585477e2.png)

# Using Grafana's time filter

Once you setup the query, you may find that your are always querying recent data in the graphs irrespective of the time range chosen in Grafana. To solve this issue, we need to pass the Grafana's dynamic time range in the url. As per the [ThingSpeak API specification](https://community.thingspeak.com/documentation%20.../api/), you need to pass the start and end time to the url as query string params. To do that, I am appending `?start=${__from:date:YYYY-MM-DD HH:NN:SS}&end=${__to:date:YYYY-MM-DD HH:NN:SS}` to the url. You can use any variable as specified in the [grafana doc](https://grafana.com/docs/grafana/latest/variables/variable-types/global-variables/)

My full url now becomes `https://thingspeak.com/channels/38629/feed.json?start=${__from:date:YYYY-MM-DD HH:NN:SS}&end=${__to:date:YYYY-MM-DD HH:NN:SS}`. 

![image](https://user-images.githubusercontent.com/153843/108482298-392ce500-7291-11eb-9137-888fc4b3515b.png)

# Long time range

Typically for up to 2 hour you can get all the metrics without any aggregation. If you are looking for more long time range, you may need to pass aggregation filter to your url. Here in the below example, I am looking for data worth of recent 180days. So i am pulling the aggrgated data using the url `https://thingspeak.com/channels/38629/feed.json?start=${__from:date:YYYY-MM-DD HH:NN:SS}&end=${__to:date:YYYY-MM-DD HH:NN:SS}&average=1440` the key to look at here is **average=1440** where it specify to take average of each 1440 minutes/1day. You can specify more grangular time aggragation. As per the document, 10, 15, 20, 30, 60, 240, 720, 1440 are the valid timerange in minutes. Also instead average  you can use sum/timescale/median functions.

![image](https://user-images.githubusercontent.com/153843/108484741-0cc69800-7294-11eb-956b-8dfb74123301.png)

**Comment by labodj** [comment]
Hi, thanks for this, it helped me with the config.
I found the time range queries to be bugged, grafana docs says that you can't put : character there.

I solved it like so:


`&start=${__from:date:YYYY-MM-DD}%20${__from:date:HH}:${__from:date:mm}:${__from:date:ss}&end=${__to:date:YYYY-MM-DD}%20${__to:date:HH}:${__to:date:mm}:${__to:date:ss}`

**Reply by gdiana** [reply]
This is a good solution and I would say the only solution. Pity GRAFANA do not simply document this.

**Comment by bmkia13** [comment]
Hi, will the plugin work in raspberry pi grafana?

**Reply by yesoreyeram** [reply]
Give a try. Should work.

**Reply by bmkia13** [reply]
okok thanks

**Reply by bmkia13** [reply]
Ha fail to install the plugin in raspberry pi, try sudo apt-get install yesoreyeram-infinity-datasource, wrong?

**Reply by bmkia13** [reply]
this work...sudo grafana-cli --pluginUrl https://github.com/yesoreyeram/grafana-infinity-datasource/releases/download/v0.6.0-alpha2/yesoreyeram-infinity-datasource-0.6.0.zip plugins install yesoreyeram-infinity-datasource

**Reply by bmkia13** [reply]
Wow, what happen??!!, Unexpected error?? Windows 10
also, all my panel show unexpected error , all panel become fail to fetch suddenly

**Reply by bmkia13** [reply]
I think is because my data cloud is thingspeak? i try to use the thingspeak json from example, same thing happen as well

**Comment by gdiana** [comment]
Ok will try and get back. 👍

Kind regards


Gregory Diana
_______________________________________


22 Glen Grove
Glenwood
Durban 4001

Phone(h&w):031-2616761 (h)
*Fax2Email: 086 474 3450*
Cell: 0784515277
E-mail: ***@***.***
E-mail: ***@***.*** ***@***.***>
Skype: gregory.diana
Print: ***@***.***

**Reply by gdiana** [reply]
Yip it all worked ok.

**Comment by techops01** [comment]
It's not working.
I'm installed infinity grafana-cli plugins install yesoreyeram-infinity-datasource
but how to add data to infinitydb from pyhton3 parser?

**Reply by gdiana** [reply]
In what data format do you want to add? XML,JSON,CSV ?

Your python3 parser must simply provide output in any one these formats

see here for details  https://github.com/yesoreyeram/grafana-infinity-datasource

**Comment by techops01** [comment]
I need details

**Comment by techops01** [comment]
JSON
My parser output in this format on y server, but l not understand how? How
it works?

вт, 2 нояб. 2021 г., 22:01 gdiana ***@***.***>:

**Comment by techops01** [comment]
Your docker-compose file not install infinitydb on default but when I manual install infinitydb but how INPUT DATA from parser in JSON to this db??!

**Reply by gdiana** [reply]
If you look further up you will see that when you open a panel, you must select the data source Then you must select json as the input and then you can input the json either as a URL or cut and paste direct
